### ISSUE TITLE
Failing std::pow instantiation

### ISSUE THREAD

**Issue by dschwen** [roystgnr/MetaPhysicL#37]
Instantiation of `std::pow` with a DualNumber base and an immediate double constant exponent fails with `no matching function for call to 'if_else'`

See long error below. (@lindsayad have you encountered this? 

```
/Users/schwd/Programs/libmesh/installed/include/metaphysicl/dualnumber.h:415:795: error: no matching function for call to 'if_else'
template <typename T, typename D, typename T2, typename D2> inline typename CompareTypes<DualNumber<T,D>,DualNumber<T2,D2> >::supertype pow (const DualNumber<T,D>& a, const DualNumber<T2,D2>& b) { typedef typename CompareTypes<T,T2>::supertype TS; typedef typename CompareTypes<DualNumber<T,D>,DualNumber<T2,D2> >::supertype type; TS funcval = std::pow(a.value(), b.value()); return type(funcval, funcval * (b.value() * a.derivatives() / a.value() + MetaPhysicL::if_else(b.derivatives(), b.derivatives() * std::log(a.value()), b.derivatives()))); } template <typename T, typename D> inline DualNumber<T,D> pow (const DualNumber<T,D>& a, const DualNumber<T,D>& b) { T funcval = std::pow(a.value(), b.value()); return DualNumber<T,D>(funcval, funcval * (b.value() * a.derivatives() / a.value() + MetaPhysicL::if_else(b.derivatives(), b.derivatives() * std::log(a.value()), b.derivatives()))); } template <typename T, typename T2, typename D> inline typename CompareTypes<DualNumber<T2,D>,T,true>::supertype pow (const T& a, const DualNumber<T2,D>& b) { typedef typename CompareTypes<DualNumber<T2,D>,T,true>::supertype type; type newa(a); return std::pow(newa, b); } template <typename T, typename T2, typename D> inline typename CompareTypes<DualNumber<T,D>,T2>::supertype pow (const DualNumber<T,D>& a, const T2& b) { typedef typename CompareTypes<DualNumber<T,D>,T2>::supertype type; type newb(b); return std::pow(a, newb); }
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          ^~~~~~~~~~~~~~~~~~~~
/Users/schwd/Programs/libmesh/installed/include/metaphysicl/dualnumber.h:415:1412: note: in instantiation of function template specialization 'std::pow<double, MetaPhysicL::NumberArray<50, double> >' requested here
template <typename T, typename D, typename T2, typename D2> inline typename CompareTypes<DualNumber<T,D>,DualNumber<T2,D2> >::supertype pow (const DualNumber<T,D>& a, const DualNumber<T2,D2>& b) { typedef typename CompareTypes<T,T2>::supertype TS; typedef typename CompareTypes<DualNumber<T,D>,DualNumber<T2,D2> >::supertype type; TS funcval = std::pow(a.value(), b.value()); return type(funcval, funcval * (b.value() * a.derivatives() / a.value() + MetaPhysicL::if_else(b.derivatives(), b.derivatives() * std::log(a.value()), b.derivatives()))); } template <typename T, typename D> inline DualNumber<T,D> pow (const DualNumber<T,D>& a, const DualNumber<T,D>& b) { T funcval = std::pow(a.value(), b.value()); return DualNumber<T,D>(funcval, funcval * (b.value() * a.derivatives() / a.value() + MetaPhysicL::if_else(b.derivatives(), b.derivatives() * std::log(a.value()), b.derivatives()))); } template <typename T, typename T2, typename D> inline typename CompareTypes<DualNumber<T2,D>,T,true>::supertype pow (const T& a, const DualNumber<T2,D>& b) { typedef typename CompareTypes<DualNumber<T2,D>,T,true>::supertype type; type newa(a); return std::pow(newa, b); } template <typename T, typename T2, typename D> inline typename CompareTypes<DualNumber<T,D>,T2>::supertype pow (const DualNumber<T,D>& a, const T2& b) { typedef typename CompareTypes<DualNumber<T,D>,T2>::supertype type; type newb(b); return std::pow(a, newb); }
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   ^
/Users/schwd/Programs/moose/modules/tensor_mechanics/src/materials/ADComputeFiniteStrain.C:80:26: note: in instantiation of function template specialization 'std::pow<double, double, MetaPhysicL::NumberArray<50, double> >' requested here
      _Fhat[_qp] *= std::pow(ave_Fhat.det() / _Fhat[_qp].det(), 1.0 / 3.0);
```

**Comment by dschwen** [comment]
Simple code example (using the MOOSE typedefs)

```
  ADReal a;
  ADReal b;
  auto r1 = std::pow(a, b);
```

also fails with

```
ed/include/metaphysicl/dualnumber.h:415:795: error: no matching function for call to 'if_else'
template <typename T, typename D, typename T2, typename D2> inline typename CompareTypes<DualNumber<T,D>,DualNumber<T2,D2> >::supertype pow (const DualNumber<T,D>& a, const DualNumber<T2,D2>& b) { typedef typename CompareTypes<T,T2>::supertype TS; typedef typename CompareTypes<DualNumber<T,D>,DualNumber<T2,D2> >::supertype type; TS funcval = std::pow(a.value(), b.value()); return type(funcval, funcval * (b.value() * a.derivatives() / a.value() + MetaPhysicL::if_else(b.derivatives(), b.derivatives() * std::log(a.value()), b.derivatives()))); } template <typename T, typename D> inline DualNumber<T,D> pow (const DualNumber<T,D>& a, const DualNumber<T,D>& b) { T funcval = std::pow(a.value(), b.value()); return DualNumber<T,D>(funcval, funcval * (b.value() * a.derivatives() / a.value() + MetaPhysicL::if_else(b.derivatives(), b.derivatives() * std::log(a.value()), b.derivatives()))); } template <typename T, typename T2, typename D> inline typename CompareTypes<DualNumber<T2,D>,T,true>::supertype pow (const T& a, const DualNumber<T2,D>& b) { typedef typename CompareTypes<DualNumber<T2,D>,T,true>::supertype type; type newa(a); return std::pow(newa, b); } template <typename T, typename T2, typename D> inline typename CompareTypes<DualNumber<T,D>,T2>::supertype pow (const DualNumber<T,D>& a, const T2& b) { typedef typename CompareTypes<DualNumber<T,D>,T2>::supertype type; type newb(b); return std::pow(a, newb); }
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          ^~~~~~~~~~~~~~~~~~~~
/Users/schwd/Programs/moose/modules/tensor_mechanics/src/materials/ComputeFiniteStrain.C:44:18: note: in instantiation of function template specialization 'std::pow<double, MetaPhysicL::NumberArray<50, double> >' requested here
  auto r1 = std::pow(a, b);
```

What ever happened to SFINAE?!!!

**Comment by dschwen** [comment]
With

```
```

I get 

```
/Users/schwd/Programs/libmesh/installed/include/metaphysicl/dualnumber.h:32:
/Users/schwd/Programs/libmesh/installed/include/metaphysicl/dualnumber_decl.h:143:40: error: no viable conversion from returned value of type 'const MetaPhysicL::NumberArray<50, double>' to function return type 'double'
  static T value(const T2& v) { return v; }
                                       ^
/Users/schwd/Programs/libmesh/installed/include/metaphysicl/dualnumber.h:98:37: note: in instantiation of function template specialization 'MetaPhysicL::DualNumberConstructor<double, MetaPhysicL::NumberArray<50, double> >::value<MetaPhysicL::NumberArray<50, double> >' requested here
  _val (DualNumberConstructor<T,D>::value(val)),
```

It looks like std::pow is just hosed (it is implemented in `dualnumber.h` though)

**Comment by roystgnr** [comment]
Huh.  It looks like we don't actually have any unit tests taking ```pow(DualNumber,double)```, but we do have a swath of ```pow(NumberArray<DualNumber>, double```, which should use the former under the hood.  Can you add something in MetaPhysicL/test/ that replicates the problem there?  I'll try to do the same.

**Comment by lindsayad** [comment]
This will currently fail if `dualnumber` is included before `numberarray`...

**Comment by dschwen** [comment]
Found two cases of "wrong" include order in MOOSE and swapping it fixes it.

**Comment by lindsayad** [comment]
@roystgnr Perhaps we should add `numberarray_decl` and `numbervector_decl` and put these at the top of `dualnumber`. Or I guess manually put some of these declarations at the top...

**Comment by roystgnr** [comment]
Was this resolved by #38?

**Comment by lindsayad** [comment]
Yep!
